### PR TITLE
WIP: boot: refactor into package and support multiboot

### PR DESCRIPTION
**Pull request by flammit**
The current commits make no changes to the existing boot cmd yet, just setting up the framework for parsing and testing.

### CODE REVIEW

File: pkg/boot/boot.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+// KexecLoad calls the appropriate kexec load routines based on the type of Entry
+func (e *Entry) KexecLoad() error {
+	switch e.Type {
+	case Multiboot:
+		// TODO: implement using kexec_load syscall
+		return syscall.ENOSYS
+	case Elf:
+		// TODO: implement using kexec_file_load syscall
```
Comment (hugelgupf) [inline pkg/boot/boot.go]: u-root has a kexec package under pkg that currently only supports kexec_file_load - you should extend that if possible.

Comment (flammit) [reply pkg/boot/boot.go]: yup - there will soon be a dependency here on pkg/kexec to use the existing file_load call, then the load when it is available.

**Comment by flammit** [comment]
no worries @hugelgupf - this probably shouldn't be merged until it's hooked into `cmd/boot` with at least a working `kexec_file_load` call, no regression on functionality (missing default logic currently), and when I've been able to test it on hardware =P

**Comment by flammit** [comment]
still WIP - waiting to test this on x230 hardware using this branch of heads (https://github.com/flammit/heads/tree/u-root) along with manually copying flashrom / libc.so / libpci.so into uinitrd/{bin,lib} to test internal reflash

File: pkg/boot/parser.go [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+		p.entry.Type = Multiboot
+	} else if strings.HasSuffix(val, ".c32") {
+		// skip this entry - not valid for kexec
+		p.state = search
+	} else {
+		p.entry.Modules = append(p.entry.Modules, NewModule(val, nil))
+	}
+}
+
+func (p *parser) parseSyslinuxAppend(val string) {
```
Comment (hugelgupf) [inline pkg/boot/parser.go]: *sigh* it looks like we're both working on a syslinux parser. #489 is working towards a pkg/pxe (and a nicer parser, once I tear apart the internals and add more tests).

I'm a bit confused as to why you have two different parsers in one structure? Grub and syslinux?

Comment (rminnich) [reply pkg/boot/parser.go]: two parsers in one struct:because the parts we care about are so similar that it does not matter: a limited incorrect thing covers almost all cases. And as you point out the ramifications of the include directive are huge since it can cover including fairly arbitrary things. And, the grub command language, which is now turing complete, is supposed to become much simpler in 2018 -- we had a discussion with phicoder about this. We proposed JSON or YAML and that seemed acceptable. 

So, for now, dead simple and not very capable parser is a reasonable solution; the fully capable parser for grub is a really big thing and will take a long time to get right. Something that's clearly limited is almost always better than something that claims to be fully capable but is not. 

A huge amount of syslinux and grub config involves menus and colors and selections and a command line -- all because we're the first thing running from the BIOS. How much of that do we really need given we are invoking a boot command from a shell under a Linux kernel? Almost none of it. We just need to know the names of the boot choices and the kernel/initramfs/cmdline that go with them. So, sure, write a full parser but be mindful of how much of it we need vs. how much of it we actually implement.

Comment (hugelgupf) [reply pkg/boot/parser.go]: I didnt say anything about a full parser... Neither of us have been working on a full one. Mine covers one more directive than this one.

We talked on slack and decided to eventually converge these two parsers, but keep working on it for now.

File: pkg/uroot/uroot.go [hunk 3]
```diff
@@ -160,6 +166,20 @@ func CreateInitramfs(opts Opts) error {
 		}
 	}
 
+	if opts.ExtraDir != "" {
```
Comment (hugelgupf) [inline pkg/uroot/uroot.go]: extrafiles should already cover this use case. Why is this necessary?

Comment (flammit) [reply pkg/uroot/uroot.go]: A couple lazy reasons: 1) the LDD routine was failing on including the musl flashrom in Heads, 2) I couldn't figure out how to get all files in a directory included in the archive relative to the root (e.g. all files in "/tmp/uinitrd-root/{etc,dev}/" included at "/{etc,dev}/".  This seemed like the cheapest / widely usable option.

 I suppose I could use the base cpio feature instead.  Am I missing something obvious?

Comment (rminnich) [reply pkg/uroot/uroot.go]: so, let's try to fix our tools if we can rather than providing workarounds in our tools for problems in our other tools :-)
You mention two issues here. Can you give us commandline examples so we can fix them?

File: pkg/boot/parser.go (hunk 2 above)
Comment (rminnich) [reply pkg/boot/parser.go]: got it. I got confused.

File: pkg/uroot/uroot.go (hunk 3 above)
Comment (flammit) [reply pkg/uroot/uroot.go]: done

**Comment by flammit** [comment]
PTAL - I didn't want to touch the original boot command until @vejmarie had a chance to test to make sure I didn't break anything. Also I dropped the MBR check (it didn't seem to be working). It'd also be nice to have better fstype detection eventually.

**Comment by vejmarie** [comment]
@flammit thanks for the work ! I am back in Europe and will have a look to it during the week for sure. I am just polishing a few things regarding virtualization support and will clearly test all of that.

**Comment by vejmarie** [comment]
I just tried to test it, but I am unable to boot a standard ubuntu Xenial with the command boot2. I might be wrond on some parameter. Here is the try I made
`
% boot2
EXT4-fs (sda): VFS: Can't find ext4 filesystem
EXT4-fs (sda): VFS: Can't find ext4 filesystem
EXT4-fs (sda): VFS: Can't find ext4 filesystem
FAT-fs (sda): invalid media value (0x90)
FAT-fs (sda): Can't find a valid FAT filesystem
FAT-fs (sda): invalid media value (0x90)
FAT-fs (sda): Can't find a valid FAT filesystem
2018/02/06 12:03:11 Config #0: path: /tmp/boot-391618391/boot/grub/grub.cfg
2018/02/06 12:03:11 Config #1: path: /tmp/boot-391618391/isolinux/isolinux.cfg
2018/02/06 12:03:11 Multiple configs found - must specify a config index
wait: exit status 1
% boot2 -help
Usage of boot2:
  -append string
    	Additional kernel params
  -c string
    	Config index
  -d string
    	Device index
  -dev string
    	Device glob (default "/sys/class/block/*")
  -dryrun
    	Only print out kexec commands
  -n string
    	Entry index
  -v	Print debug messages
wait: exit status 2
% boot2 -c 0
2018/02/06 12:04:06 No entry specified
wait: exit status 1
% boot2 -c 0 -n 0
2018/02/06 12:04:26 Invalid entry index:0
wait: exit status 1
% boot2 -c 0 -n 1
2018/02/06 12:04:33 Invalid entry index:1
wait: exit status 1
% boot2 -c 0 -n 1 -v
2018/02/06 12:04:41 Got devices: []*boot.Device{(*boot.Device)(0xc420056550)}
2018/02/06 12:04:41 Got configs: []*boot.Config{(*boot.Config)(0xc420043040), (*boot.Config)(0xc4200430c0)}
2018/02/06 12:04:41 Got entries: []boot.Entry(nil)
2018/02/06 12:04:41 Invalid entry index:1
wait: exit status 1
% 
`
Then if I use boot command (original one) I can't boot without running a reset of the server. looks like the new command is mounting the file system and doesn't umount it first ?
`
% boot -dryrun=false
2018/02/06 12:12:42 kexec on /tmp/u-root-boot836283797/dev/sda failed: Entry default not found in config file
2018/02/06 12:12:42 Kexec Reboot /tmp/u-root-boot836283797/dev/sda failed, sys_reboot(..., kexec) = invalid argument. Sorry
2018/02/06 12:12:42 Sorry no bootable device found
wait: exit status 1
% 
`
after a server reset the boot command works (the original) and is able to boot the server.

My grub.cfg entry looks like
`
menuentry 'Ubuntu' --class ubuntu --class gnu-linux --class gnu --class os $menuu
entry_id_option 'gnulinux-simple-7be3beb0-b1f5-4832-84c0-c621f526fe06' {
        recordfail
        load_video
        gfxmode $linux_gfx_mode
        insmod gzio
        if [ x$grub_platform = xxen ]; then insmod xzio; insmod lzopio; fi
        insmod part_gpt
        insmod ext2
        set root='hd0,gpt2'
        if [ x$feature_platform_search_hint = xy ]; then
          search --no-floppy --fs-uuid --set=root --hint-bios=hd0,gpt2 --hint-eff
i=hd0,gpt2 --hint-baremetal=ahci0,gpt2  7be3beb0-b1f5-4832-84c0-c621f526fe06
        else
          search --no-floppy --fs-uuid --set=root 7be3beb0-b1f5-4832-84c0-c621f55
26fe06
        fi
        linux   /boot/vmlinuz-4.4.0-109-generic root=UUID=7be3beb0-b1f5-4832-84cc
0-c621f526fe06 ro  console=tty0 console=ttyS4,57600 console=ttyS0,57600 vt.hando
ff=7
        initrd  /boot/initrd.img-4.4.0-109-generic
}
submenu 'Advanced options for Ubuntu' $menuentry_id_option 'gnulinux-advanced-7bb
e3beb0-b1f5-4832-84c0-c621f526fe06' {
`

**Comment by flammit** [comment]
Apologies for the long delay here... I'll rebase and debug the issues @vejmarie had now that it's easier for me to test with the heads/linuxboot config.

**Comment by insomniacslk** [comment]
Hi @flammit, I'm also looking at local booting. I could boot a debian and an ubuntu image without problems on Qemu, though I think that @vejmarie 's tests are on real hardware and are definitely more indicative.

A quick look at the code - you use `log.Fatal` throughout the code, which breaks the `defer` mechanism. Basically when called it prints the error message and exits immediately, so defers are not run. This seems exactly what @vejmarie is seeing - disks not being unmounted because the deferred `cleanDevices` is not run.

Can you convert all the `log.Fatal` calls to return errors instead, and let the error bubble up to the `main` where it can be handled safely?

**Comment by flammit** [comment]
@insomniacslk wilco - I'm going to rebase and test this again in qemu/x230 this weekend.  I had no problems with debian/ubuntu myself and I don't have a linuxboot hardware to test locally here, so I'll do what I can and add the debugging.

**Comment by insomniacslk** [comment]
thanks @flammit . We are all gathered in a room today, so I will probably see what's to be fixed here and report back to you to update the PR

**Comment by insomniacslk** [comment]
I committed the change to replace log.Fatal with error handling, this should resolve the cleanup issue that @vejmarie was seeing.

cc @hugelgupf

**Comment by insomniacslk** [comment]
I guess we need @vejmarie 's help to test this on real hardware. I will also try to build and test an image for a wedge in the next days

**Comment by hugelgupf** [comment]
Rebased to master, moved to `pkg/diskboot`. Going to incorporate `pkg/boot` into this. (Entry -> boot.OSImage)

**Comment by insomniacslk** [comment]
just as an FYI you may also want to look at https://github.com/insomniacslk/systemboot/tree/master/localboot . I was planning to merge it in u-root (or its functionalities into the local booting package)

**Comment by vejmarie** [comment]
Hey guys, why forking new project to develop specific command and not building a new branch directly in u-root while waiting a merge in master. @insomniacslk your work seems really great and was needed by the way, but it might be easier to follow it with multiple "small" commit into original tree ;)

**Comment by insomniacslk** [comment]
@vejmarie I needed to fork for some internal need, but I am happy to merge upstream. In the end I don't want to maintain stuff myself , there's a whole community of great engineers out there :)

**Comment by hugelgupf** [comment]
We're all in agreement that we need to converge on one thing in the u-root tree, I think :)

**Comment by insomniacslk** [comment]
@hugelgupf @vejmarie how would you see merging the `localboot` mentioned above into u-root? No rush on my side, as @zaolin and I are still working on some feature that we'd like to iterate fast on, then everything can go upstream

**Comment by hugelgupf** [comment]
There are one or two things I want to do before we add more features - I want to use boot.OSImage instead of Entry, and then move to the kexec stuff over there. Also, this is the second syslinux parser in u-root - the pxe parser is basically the same thing. I would like to share that.

But this doesn't have to happen in that order. I'm going to just merge this as is when I get back.
